akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -113,41 +114,46 @@ private[http] object StreamUtils {
     Flow[ByteString].transform(() ⇒ transformer).named("sliceBytes")
   }
 
-  def limitByteChunksStage(maxBytesPerChunk: Int): PushPullStage[ByteString, ByteString] =
-    new StatefulStage[ByteString, ByteString] {
-      def initial = WaitingForData
+  def limitByteChunksStage(maxBytesPerChunk: Int): GraphStage[FlowShape[ByteString, ByteString]] =
+    new SimpleLinearGraphStage[ByteString] {
+      override def initialAttributes = Attributes.name("limitByteChunksStage")
+      var remaining = ByteString.empty
 
-      case object WaitingForData extends State {
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective =
-          if (elem.size <= maxBytesPerChunk) ctx.push(elem)
-          else {
-            become(DeliveringData(elem.drop(maxBytesPerChunk)))
-            ctx.push(elem.take(maxBytesPerChunk))
-          }
-      }
-
-      case class DeliveringData(remaining: ByteString) extends State {
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective =
-          throw new IllegalStateException("Not expecting data")
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
 
-        override def onPull(ctx: Context[ByteString]): SyncDirective = {
-          val toPush = remaining.take(maxBytesPerChunk)
-          val toKeep = remaining.drop(maxBytesPerChunk)
-
-          become {
-            if (toKeep.isEmpty) WaitingForData
-            else DeliveringData(toKeep)
+        def splitAndPush(elem: ByteString): Unit = {
+          val (toSend, toRemain) = elem.splitAt(maxBytesPerChunk)
```

Comment:
it's one line instead of two. still uses the same two calls slice inside ByteSring